zipkin-server/src/main/java/zipkin2/server/internal/ZipkinQueryApiV2.java

```diff
@@ -71,116 +64,109 @@
+    @Override public int sizeInBytes(String value) {
+      return Buffer.utf8SizeInBytes(value) + 2; // quotes
+    }
 
-  /**
-   * We cache tag values to minimize the number of requests made to the storage backend. The tag
-   * values doesn't change frequently and cache expires in 5 minutes
-   *
-   */
-  ResponseEntity<List<String>> maybeCacheAutocompleteValues(List<String> values) {
-    ResponseEntity.BodyBuilder response = ResponseEntity.ok();
-    if (values.size() > 3) {
-      response.cacheControl(CacheControl.maxAge(namesMaxAge, TimeUnit.SECONDS).mustRevalidate());
+    @Override public void write(String value, Buffer buffer) {
+      buffer.writeByte('"').writeUtf8(value).writeByte('"');
     }
-    return response.body(values);
-  }
+  };
 
-  @ExceptionHandler(TraceNotFoundException.class)
-  @ResponseStatus(HttpStatus.NOT_FOUND)
-  public void notFound() {}
+  @Get("/api/v2/autocompleteKeys")
+  public AggregatedHttpMessage getAutocompleteKeys() {
+    return maybeCacheNames(true, autocompleteKeys);
+  }
 
-  static class TraceNotFoundException extends RuntimeException {
-    TraceNotFoundException(String traceIdHex) {
-      super("Cannot find trace " + traceIdHex);
-    }
+  @Get("/api/v2/autocompleteValues")
+  public AggregatedHttpMessage getAutocompleteValues(@Param("key") String key) throws IOException {
+    List<String> values = storage.autocompleteTags().getValues(key).execute();
+    return maybeCacheNames(values.size() > 3, values);
   }
 
   /**
-   * We cache names if there are more than 3 services. This helps people getting started: if we
-   * cache empty results, users have more questions. We assume caching becomes a concern when zipkin
-   * is in active use, and active use usually implies more than 3 services.
+   * We cache names if there are more than 3 names. This helps people getting started: if we cache
+   * empty results, users have more questions. We assume caching becomes a concern when zipkin is in
+   * active use, and active use usually implies more than 3 services.
    */
-  ResponseEntity<List<String>> maybeCacheNames(List<String> names) {
-    ResponseEntity.BodyBuilder response = ResponseEntity.ok();
-    if (serviceCount > 3) {
-      response.cacheControl(CacheControl.maxAge(namesMaxAge, TimeUnit.SECONDS).mustRevalidate());
+  AggregatedHttpMessage maybeCacheNames(boolean shouldCacheControl, List<String> values) {
+    byte[] body = JsonCodec.writeList(QUOTED_STRING_WRITER, values);
+    HttpHeaders headers = HttpHeaders.of(200)
+      .contentType(MediaType.JSON)
+      .setInt(HttpHeaderNames.CONTENT_LENGTH, body.length);
+    if (shouldCacheControl) {
+      headers = headers.add(
+        HttpHeaderNames.CACHE_CONTROL,
+        CacheControl.maxAge(namesMaxAge, TimeUnit.SECONDS).mustRevalidate().getHeaderValue()
+      );
     }
-    return response.body(names);
+    return AggregatedHttpMessage.of(headers, HttpData.of(body));
```

Comment:
not sure it must be `AggregatedHttpMessage` or we could just use HttpResponse.of() here ? cc @hyangtack

Reply:
Yep. A user can use `HttpResponse.of(...)`. `AggregatedHttpMessage` will be translated into `HttpResponse` anyway.

Reply:
I think HttpReaponse will be on the event loop, not blocking executor. There's some blocking here (though easily made async which we should at some point)